pkg/http/handler/timeout.go

```diff
@@ -75,25 +78,55 @@ func (h *timeToFirstByteTimeoutHandler) ServeHTTP(w http.ResponseWriter, r *http
 		h.handler.ServeHTTP(tw, r.WithContext(ctx))
 	}()
 
-	timeout := getTimer(h.timeout)
-	var timeoutDrained bool
-	defer func() { putTimer(timeout, timeoutDrained) }()
+	firstByteTimeout := getTimer(h.firstByteTimeout)
+	var firstByteTimeoutDrained bool
+	defer func() {
+		putTimer(firstByteTimeout, firstByteTimeoutDrained)
+	}()
+
+	var idleTimeout *time.Timer
+	var idleTimeoutDrained bool
+	if h.idleTimeout > 0 {
+		idleTimeout = getTimer(h.idleTimeout)
+		defer func() {
+			putTimer(idleTimeout, idleTimeoutDrained)
+		}()
+	}
+
 	for {
 		select {
 		case p, ok := <-done:
 			if ok {
 				panic(p)
 			}
 			return
-		case <-timeout.C:
-			timeoutDrained = true
-			if tw.timeoutAndWriteError(h.body) {
+		case <-firstByteTimeout.C:
+			firstByteTimeoutDrained = true
+			if tw.tryFirstByteTimeoutAndWriteError(h.body) {
 				return
 			}
+		case <-timerChannel(idleTimeout):
```

Comment:
would it be possible to drop this method and instead initialise `idleTimeoutCh` to either nil (using the fact a read from a nil channel blocks) or idleTimeout.C above?

i.e.

~~~~
var idleTimeoutCh <-chan time.Time
if idleTimeout != nil {
   idleTimeoutCh = idleTimeout.C
}
~~~~

that seems a bit clearer, and avoids checking for nil every iteration of the loop

Reply:
Ohhh I forgot about that. That'd indeed be even better than my suggestions.

Reply:
Great idea! Fixed